crates/ra_assists/src/add_missing_impl_members.rs

```diff
@@ -0,0 +1,279 @@
+            buf.push_str(" { unimplemented!() }")
+        } else {
+            child.text().push_to(&mut buf);
+        }
+    }
+
+    buf.trim_end().to_string()
+}
+
+pub(crate) fn add_missing_impl_members(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let node = ctx.covering_node();
+    let impl_node = node.ancestors().find_map(ast::ImplBlock::cast)?;
+    let impl_item_list = impl_node.item_list()?;
+    // Don't offer the assist when cursor is at the end, outside the block itself.
+    let cursor_range = TextRange::from_to(node.range().end(), node.range().end());
+    if node.range().end() == impl_node.syntax().range().end() {
+        return None;
+    }
+
+    let trait_def = {
+        let position = FilePosition { file_id: ctx.frange.file_id, offset: cursor_range.end() };
+        let resolver = hir::source_binder::resolver_for_position(ctx.db, position);
+
+        resolve_target_trait_def(ctx.db, &resolver, impl_node)?
+    };
+
+    let missing_fns: Vec<_> = {
+        let fn_def_opt = |kind| if let ImplItemKind::FnDef(def) = kind { Some(def) } else { None };
+        let def_name = |def| -> Option<&SmolStr> { FnDef::name(def).map(ast::Name::text) };
+
+        let trait_items =
+            trait_def.syntax().descendants().find_map(ast::ItemList::cast)?.impl_items();
+        let impl_items = impl_item_list.impl_items();
+
+        let trait_fns = trait_items.map(ImplItem::kind).filter_map(fn_def_opt).collect::<Vec<_>>();
+        let impl_fns = impl_items.map(ImplItem::kind).filter_map(fn_def_opt).collect::<Vec<_>>();
+
+        trait_fns
+            .into_iter()
+            .filter(|t| def_name(t).is_some())
+            .filter(|t| impl_fns.iter().all(|i| def_name(i) != def_name(t)))
+            .collect()
+    };
+    if missing_fns.is_empty() {
+        return None;
+    }
+
+    ctx.add_action(AssistId("add_impl_missing_members"), "add missing impl members", |edit| {
+        let (parent_indent, indent) = {
+            // FIXME: Find a way to get the indent already used in the file.
+            // Now, we copy the indent of first item or indent with 4 spaces relative to impl block
+            const DEFAULT_INDENT: &str = "    ";
+            let first_item = impl_item_list.impl_items().next();
+            let first_item_indent =
+                first_item.and_then(|i| leading_indent(i.syntax())).map(ToOwned::to_owned);
+            let impl_block_indent = leading_indent(impl_node.syntax()).unwrap_or_default();
+
+            (
+                impl_block_indent.to_owned(),
+                first_item_indent.unwrap_or_else(|| impl_block_indent.to_owned() + DEFAULT_INDENT),
+            )
+        };
+
+        let changed_range = {
+            let last_whitespace = impl_item_list.syntax().children();
+            let last_whitespace = last_whitespace.filter_map(ast::Whitespace::cast).last();
+
+            last_whitespace.map(|w| w.syntax().range()).unwrap_or(cursor_range)
```

Comment:
why do we want to *replace* some whitespace? Can we use a simpler logic like "add new items before the `}`"? We'll need to get `}` before `add_action` and return `None` if it isn't there.

Reply:
@matklad This is to provide some form of formatting (e.g. `test_add_missing_impl_members`):
In the case below
```rust
impl Foo for S {
    fn bar(&self) {}
    <|>
}
```

otherwise we'd insert
```rust
impl Foo for S {
    fn bar(&self) {}

    fn foo(&self) { unimplemented!() }
    fn baz(&self) { unimplemented!() }<|>
}
```

so the general approach of replacing the last whitespace between `}` and the last item formats it better (and that's what also was more intuitive for me). Do you mind if leave it as-is?

Reply:
Yeah, let's keep it this way then! The real fix is to use formatter of course, when we implement that!